### PR TITLE
E2E Tests: Verify the expected theme is active.

### PR DESCRIPTION
**Pull request by epiqueras**
## Description

This PR adds a check before all E2E tests are ran that verifies that the theme the tests expect is active.

Sometimes, contributors run tests locally with different themes and are confused as to why they fail. This check will surface those situations. See [this comment](https://github.com/WordPress/gutenberg/pull/17917#issuecomment-557596676) for an example.

## How has this been tested?

It was verified that running the E2E tests with a theme other than the currently expected one, twentytwenty, throws a descriptive error before all tests are run.

## Checklist:

- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

**Comment by mcsf** [comment]
Hm, the thing with a check added in the global config like this is that we suddenly require Twenty Twenty for any tests regardless of their needs. I think there are legitimate cases where you’d want to run given test suites against a site with a different active theme. Sure, the global config may be bypassed, but how practical is that?

This is why I was speculating in the other issue about a suite-specific theme constraint (and hence the importance of performance).

**Comment by epiqueras** [comment]
>Hm, the thing with a check added in the global config like this is that we suddenly require Twenty Twenty for any tests regardless of their needs.

We already do, implicitly. All of our tests expect the latest theme and our CI/CD pulls from `trunk`. I also only added it to the `e2e` package. I didn't add it to the `test-e2e` script in `@wordpress/scripts` precisely for these reasons.

File: packages/e2e-tests/config/setup-test-framework.js [hunk 1]
```diff
@@ -50,6 +51,19 @@ async function setupBrowser() {
 	await setBrowserViewport( 'large' );
 }
 
+/**
+ * Verifies that the expected theme for the tests is active.
+ */
+async function verifyActiveTheme() {
+	await page.goto(
+		createURL( '' )
+	);
```
Comment (mcsf) [inline packages/e2e-tests/config/setup-test-framework.js]: What does this ensure, and can we achieve the same perhaps with a more semantic use of Puppeteer's API?

**Comment by mcsf** [comment]
> I also only added it to the `e2e` package. I didn't add it to the `test-e2e` script in `@wordpress/scripts` precisely for these reasons.

I'm not sure I understand the distinction. Doesn't `test-e2e` load the [config](https://github.com/WordPress/gutenberg/blob/34848321af26f11b6fa9eea81dd2772ce6198810/packages/e2e-tests/jest.config.js#L7) and set up the framework [based on](https://github.com/WordPress/gutenberg/blob/34848321af26f11b6fa9eea81dd2772ce6198810/package.json#L192) `e2e`?

File: packages/e2e-tests/config/setup-test-framework.js (hunk 1 above)
Comment (epiqueras) [reply packages/e2e-tests/config/setup-test-framework.js]: It takes you to the home page in the front end of the site so that the theme's stylesheets will be loaded.

>a more semantic use of Puppeteer's API?

What were you thinking?

**Comment by epiqueras** [comment]
>I'm not sure I understand the distinction. Doesn't test-e2e load the config and set up the framework based on e2e?

Yes, this means other projects using `test-e2e` won't be affected. I guess you are concerned about projects using the Gutenberg test suites to test integration with their plugins. I can see how a more granular use of `verifyActiveTheme` would be useful there if they only run a subset of the test suites.

How do you feel about tracking down which test suites rely on a specific theme and adding `verifyActiveTheme` there as granularly as possible?

File: packages/e2e-test-utils/src/verify-active-theme.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+/**
+ * Verifies that the expected theme for the tests is active.
+ *
+ * @param {string} theme The expected theme's slug.
+ */
+export async function verifyActiveTheme( theme ) {
+	await page.goto( createURL( '' ) );
+	if ( ( await page.$( `#${ theme }-style-css` ) ) === null ) {
+		process.exit( `Test suite must be run using the ${ theme } theme.` );
```
Comment (epiqueras) [inline packages/e2e-test-utils/src/verify-active-theme.js]: We have to use `process.exit`, because of [this bug in Jest](https://github.com/facebook/jest/issues/6695).

**Comment by gziolo** [comment]
How feasible it would be to update the existing test to be theme independent? Can we customize the theme by setting our own colors and fonts which override theme defaults? Alternatively, can we override CSS or element styles in places where it becomes theme-specific and fails tests?

File: packages/e2e-tests/config/setup-test-framework.js (hunk 1 above)
Comment (mcsf) [reply packages/e2e-tests/config/setup-test-framework.js]: I was thrown off by my own expectation of not having to navigate anywhere. Indeed, with a one-off requirement specially made for Twenty Twenty it would be enough to look for `#twentytwenty-block-editor-styles-css` in the editor itself.

What you're doing is more generic, since the existence of `#foo-style-css` in the front end is guaranteed by WordPress itself, not the theme implementation. There's no such guarantee for the back end, alas, as editor styles are optional.

In short: disregard my comment about anything more semantic.

**Comment by epiqueras** [comment]
>How feasible it would be to update the existing tests to be theme independent? Can we customize the theme by setting our own colors and fonts which override theme defaults? Alternatively, can we override CSS or element styles in places where it becomes theme-specific and fails tests?

I think that is a much better approach. I tried it here: #18761, and it seems to work well. Should we close this and merge that?

**Comment by mcsf** [comment]
Yes, that sounds much better, and #18761 already looks promising.

**Comment by epiqueras** [comment]
Closing this then.
